File: src/banner/renderer.py
"""SVG rendering for translated banners."""

import math
from xml.sax.saxutils import escape

from PIL import Image

from .models import TranslatedText
from .text_style import image_to_data_uri


def render_svg(
    background_image: Image.Image,
    translated_texts: list[TranslatedText],
) -> str:
    """
    Render translated texts as an SVG overlay on an image.

    Args:
        background_image: The inpainted background image
        translated_texts: List of translated text elements with positioning

    Returns:
        SVG string with the rendered banner
    """
    width = background_image.width
    height = background_image.height

    svg_parts = [
        f'<svg width="{width}" height="{height}" xmlns="http://www.w3.org/2000/svg">',
        f'<image href="{image_to_data_uri(background_image)}" width="{width}" height="{height}" />',
    ]

    for text in translated_texts:
        box = text.bounding_box
        style = text.style
        center = box.center
        box_width, _ = box.dimensions
        angle_deg = math.degrees(box.angle)

        text_elem = (
            f'<text x="{center.x}" y="{center.y}" '
            f'fill="{style.foreground_color}" '
            f'stroke="{style.border_color}" stroke-width="2" '
            f'font-weight="{style.font_weight}" '
            f'font-family="{text.font_family}" '
            f'dominant-baseline="middle" '
            f'font-size="{style.font_size}" '
            f'text-anchor="middle" '
            f'paint-order="stroke fill" '
            f'textLength="{box_width}" '
            f'lengthAdjust="spacingAndGlyphs" '
            f'transform="rotate({angle_deg})" '
            f'transform-origin="center">'
            f"{escape(text.translated)}</text>"
        )
        svg_parts.append(text_elem)

    svg_parts.append("</svg>")
    return "\n".join(svg_parts)

File: workers/banner/src/renderer.ts
/**
 * SVG rendering for translated banners.
 */

import type { TranslatedText } from "./types";
import { getBoxCenter, getBoxDimensions, getBoxAngle } from "./types";

/**
 * Escape special XML characters.
 */
function escapeXml(text: string): string {
	return text
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&apos;");
}

/**
 * Render translated texts as an SVG overlay on an image.
 *
 * @param backgroundImageDataUri - Base64 data URI of the inpainted background image
 * @param width - Image width
 * @param height - Image height
 * @param translatedTexts - List of translated text elements with positioning
 * @returns SVG string with the rendered banner
 */
export function renderSvg(
	backgroundImageDataUri: string,
	width: number,
	height: number,
	translatedTexts: TranslatedText[]
): string {
	const svgParts: string[] = [
		`<svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg">`,
		`<image href="${backgroundImageDataUri}" width="${width}" height="${height}" />`,
	];

	for (const text of translatedTexts) {
		const box = text.boundingBox;
		const style = text.style;
		const center = getBoxCenter(box);
		const { width: boxWidth } = getBoxDimensions(box);
		const angleDeg = (getBoxAngle(box) * 180) / Math.PI;

		const textElem =
			`<text x="${center.x}" y="${center.y}" ` +
			`fill="${style.foregroundColor}" ` +
			`stroke="${style.borderColor}" stroke-width="2" ` +
			`font-weight="${style.fontWeight}" ` +
			`font-family="${text.fontFamily}" ` +
			`dominant-baseline="middle" ` +
			`font-size="${style.fontSize}" ` +
			`text-anchor="middle" ` +
			`paint-order="stroke fill" ` +
			`textLength="${boxWidth}" ` +
			`lengthAdjust="spacingAndGlyphs" ` +
			`transform="rotate(${angleDeg})" ` +
			`transform-origin="center">` +
			`${escapeXml(text.translated)}</text>`;

		svgParts.push(textElem);
	}

	svgParts.push("</svg>");
	return svgParts.join("\n");
}
